src/group_theory/divisible.lean

```diff
@@ -0,0 +1,385 @@
+      rw [int.of_nat_eq_coe] at hn,
+      exact_mod_cast hn, },
+    { norm_num,
+      change - ((z+1) • - (divisible_by.div _ _)) = _,
+      have := nsmul_zero_sub (divisible_by.div a (z + 1)) (z + 1),
+      rw [zero_sub, zero_sub] at this,
+      rw [this, neg_neg, divisible_by.div_cancel],
+      norm_num, },
+  end}
+
+/--
+An add_group is `ℕ`-divisible if it is `ℤ`-divisible.
+-/
+def divisible_by_nat_of_divisible_by_int [divisible_by A ℤ] :
+  divisible_by A ℕ :=
+{ div := λ a n, divisible_by.div a (n : ℤ),
+  div_zero := λ a, divisible_by.div_zero a,
+  div_cancel := λ n a hn, begin
+    have := divisible_by.div_cancel a (by exact_mod_cast hn : (n : ℤ) ≠ 0),
+    norm_num at this,
+    assumption,
+  end }
+
+end add_group
+
+namespace add_comm_group
+
+open add_monoid
+
+variables (A : Type*) [add_comm_group A]
+
+lemma smul_top_eq_top_of_divisible_by_int [divisible_by A ℤ] {n : ℤ} (hn : n ≠ 0) :
+  n • (⊤ : add_subgroup A) = ⊤ :=
+add_subgroup.map_top_of_surjective _ $ λ a, ⟨divisible_by.div a n, divisible_by.div_cancel _ hn⟩
+
+/--
+If for all `n ≠ 0 ∈ ℤ`, `n • A = A`, then `A` is divisible.
+-/
+noncomputable def divisible_by_int_of_smul_top_eq_top
+  (H : ∀ {n : ℤ} (hn : n ≠ 0), n • (⊤ : add_subgroup A) = ⊤) :
+  divisible_by A ℤ :=
+{ div := λ a n, dite (n = 0) (λ _, 0)
+    (λ hn, (show a ∈ n • (⊤ : add_subgroup A), by rw [H hn]; trivial).some),
+  div_zero := λ a, dif_pos rfl,
+  div_cancel := λ n a hn, begin
+    rw [dif_neg hn],
+    generalize_proofs h1,
+    exact h1.some_spec.2,
+  end }
+
+@[priority 100]
+instance divisible_of_char_zero {𝕜} [division_ring 𝕜] [char_zero 𝕜] : divisible_by 𝕜 ℤ :=
+{ div := λ q n, q / n,
+  div_zero := λ q, by norm_num,
+  div_cancel := λ n q hn,
+    by rw [zsmul_eq_mul, (int.cast_commute n _).eq, div_mul_cancel q (int.cast_ne_zero.mpr hn)] }
+
+section quotient
+
+variables {B : add_subgroup A} [divisible_by A ℕ]
+
+/--
+Any quotient group of a divisible group is divisible.
+-/
+noncomputable def divisible_by_quotient : divisible_by (A ⧸ B) ℕ :=
+add_monoid.divisible_by_of_smul_surj _ _ $ λ n hn x, quotient.induction_on' x $ λ a,
+  ⟨quotient.mk' (divisible_by.div a n),
+    (congr_arg _ (divisible_by.div_cancel _ hn) : quotient.mk' _ = _)⟩
+
+end quotient
+
+section hom
+
+variables {A} [divisible_by A ℕ] {B : Type*} [add_comm_group B] (f : A →+ B)
+
+/--
+If `f : A → B` is a surjective group homomorphism and `A` is divisible, then `B` is divisible.
+-/
+noncomputable def divisible_by_of_surj (hf : function.surjective f) : divisible_by B ℕ :=
+add_monoid.divisible_by_of_smul_surj _ _ $ λ n hn x,
+  ⟨f $ divisible_by.div (hf x).some n, by rw [←f.map_nsmul (divisible_by.div (hf x).some n) n,
+    divisible_by.div_cancel _ hn, (hf x).some_spec]⟩
+
+end hom
+
+
+end add_comm_group
+
+namespace group
+
+open monoid
+
+variables (A : Type*) [group A]
+
+/--
+A group is `ℤ`-rootable if it is `ℕ`-rootable.
+-/
+def rootable_by_int_of_rootable_by_nat [rootable_by A ℕ] :
```

Comment:
`to_additive` can is already almost smart enough to generate the `divisible` variant from this definition!

If you update the translation table `tr` in `src/algebra/group/to_additive.lean` to add the case:
```lean
| is_comm ("rootable" :: s) := add_comm_prefix is_comm "divisible" :: tr ff s
```
then you can do
```suggestion
@[to_additive]
def rootable_by_int_of_rootable_by_nat [rootable_by A ℕ] :
```
and everything happens automatically.

Reply:
thanks for the suggestion.